## Very serious note: Do not use any libraries except for numpy or standard math functions. Implement everything by hand on paper first, this will help you understand the whole process and get some more practice in this week's topic. Then simply code the process in this notebook (you can hard-code, or use any cleaner methodologies. Just make sure that the output is correct). If you are stuck, rewatch the lectures.

<img src="2.png" alt="Neural net cannot be found" width="700">

## Task 1. Implement the simple forward pass function and print the final output from the neurons. Error is squared error

In [1]:
#These are just helper functions you can use 
#(feel free to update them and even change the input if you deem it necessary)
import numpy as np

def SquaredError(target, x):
    output = 1/2*((target - x)**2)
    return output

def Sigmoid(x):
    output = 1 / (1 + np.exp(-x))
    return output

def ReLu(x):
    output = max(0, x)
    return output

In [2]:
def h1_in_function(vx1, vw1, vx2, vw2):
    return vx1*vw1+vx2*vw2

def o_in_function(vh, vw3, vx3, vw4):
    return vh*vw3+vx3*vw4

In [3]:
#put your main code here
x1, x2, x3 = 5, 7, 2.9
w1, w2, w3, w4 = 0.2, 0.4, 0.1, 0.3
expected_out = 0.9

In [4]:
def forward_pass(v_w1, v_w2, v_w3, v_w4):
    v_h1_in =  h1_in_function(x1, v_w1, x2, v_w2)
    v_h1_out = ReLu(v_h1_in)

    v_o_in = o_in_function(v_h1_out, v_w3, x3, v_w4)
    v_o_out = Sigmoid(v_o_in)
    return v_o_out

o_out = forward_pass(w1, w2, w3, w4)
print('Final output from the neuron:', round(o_out, 4))
e = SquaredError(expected_out, o_out)
print('Squared error:', round(e, 4)) 

Final output from the neuron: 0.7773
Squared error: 0.0075


## Task 2. Implement the derivatives of the following (they might be useful later, feel free to change the input variables as well)

In [5]:
#These are just helper functions you can use 
#(feel free to update them and even change the input if you deem it necessary)

def SquaredErrorDerivative(target, x):
    output = (target - x)*(-1)
    return output

def SigmoidDerivative(x):
    output = x*(1-x)
    return output

def ReLuDerivative(x):
    output = 1 if x>0 else 0
    return output

In [6]:
#put your main code here
def e_w3_derivative():
    v_h1_in =  h1_in_function(x1, w1, x2, w2)
    v_h1_out = ReLu(v_h1_in)
    return v_h1_out

def e_w4_derivative():
    return x3

def e_h1_derivative():
    return w3

def e_w1_derivative():
    return x1

def e_w2_derivative():
    return x2

## Task 3. Calculate the weight updates via the backward pass. Learning rate alpha is 0.1

In [7]:
#put your main code here
def backward_pass(v_expected_out, v_o_out):
    v_dw3 = SquaredErrorDerivative(v_expected_out, v_o_out) * SigmoidDerivative(v_o_out) * e_w3_derivative()
    v_dw4 = SquaredErrorDerivative(v_expected_out, v_o_out) * SigmoidDerivative(v_o_out) * e_w4_derivative()

    v_deh1 = SquaredErrorDerivative(v_expected_out, o_out) * SigmoidDerivative(v_o_out) * e_h1_derivative()
    v_h1_in = h1_in_function(x1, w1, x2, w2)
    v_dw1 = v_deh1 * ReLuDerivative(v_h1_in) * e_w1_derivative()
    v_dw2 = v_deh1 * ReLuDerivative(v_h1_in) * e_w2_derivative()
    return v_dw1, v_dw2, v_dw3, v_dw4

dw = backward_pass(expected_out, o_out)
print('dw:', dw)

dw: (-0.010619990697652465, -0.01486798697671345, -0.08071192930215873, -0.061595946046384284)


## Task 4. Calculate the new weights and update them

In [8]:
#put your main code here
def updete_weights(v_dw):
    a = 0.1
    w1x = w1 - a*v_dw[0]
    w2x = w2 - a*v_dw[1]
    w3x = w3 - a*v_dw[2]
    w4x = w4 - a*v_dw[3]
    return w1x, w2x, w3x, w4x

w1, w2, w3, w4 = updete_weights(dw)
print('w1+, w2+, w3+, w4+:', w1, w2, w3, w4)

w1+, w2+, w3+, w4+: 0.20106199906976527 0.40148679869767134 0.10807119293021587 0.30615959460463843


## Task 5. Now do a second forward pass and observe the new output. 

In [9]:
#put your main code here
o_out = forward_pass(w1, w2, w3, w4)
print('out after 1 step:', o_out)
e = SquaredError(expected_out, o_out)
print('Squared error:', round(e, 4)) 

# dw = backward_pass(expected_out, o_out)
# print('dw:', dw)

# w1, w2, w3, w4 = updete_weights(dw)
# print('w1+, w2+, w3+, w4+:', w1, w2, w3, w4)

# o_out = forward_pass(w1, w2, w3, w4)
# print('out after 2 step:', o_out)
# e = SquaredError(expected_out, o_out)
# print('Squared error:', round(e, 4)) 


out after 1 step: 0.7858740215981732
Squared error: 0.0065


As a result of minor changes in weights, we observe a significant decrease in the Squared error from 0.0075 to 0.0065